### PR TITLE
add support for plugin layers in QgsProcessingParameterMapLayer (fix #44183)

### LINKED ISSUE #44183: Need QgsProcessing::SourceType::TypePluginLayer for custom layers Processing Algorithms
Feature description.
----------------------

QGIS allows to create plugins and their own `QgsPluginLayer`/`QgsPluginLayerType` ...  
but one cannot use such layers in the processing toolbox  *(or at least I didn't find out howto)*

**In the current state, there is no way to handle a custom layer (aka `QgsPlugin`) as a `QgsProcessingAlgorithm` parameter...**

It would be awesome to get a way to *at least* add a restriction on `QgsPluginLayerTypes` to `QgsProcessingParameterMapLayer.QgsProcessingParameterLimitedDataTypes` ... and allow to emit an apropriated error/warning at runtime...

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Allow to filter plugin layers in the `QgsProcessingParameterMapLayer` and `QgsProcessingParameterMultipleLayers`.

Fixes #44183

### CODE REVIEW

**Comment by alexbruy** [comment]
Failure on Fedora is unrelated.

File: src/core/processing/qgsprocessingutils.cpp [hunk 1]
```diff
@@ -107,6 +108,29 @@ QList<QgsMeshLayer *> QgsProcessingUtils::compatibleMeshLayers( QgsProject *proj
   return layers;
 }
 
+QList<QgsPluginLayer *> QgsProcessingUtils::compatiblePluginLayers( QgsProject *project, bool sort )
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingutils.cpp:111]: This is calling out for a template based method to condense compatibleRasterLayers/compatibleMeshLayers/compatiblePluginLayers into one unified function

File: tests/src/analysis/testqgsprocessing.cpp [hunk 2]
```diff
@@ -2876,6 +2876,11 @@ void TestQgsProcessing::parameterMapLayer()
   QCOMPARE( pythonCode, QStringLiteral( "QgsProcessingParameterMapLayer('non_optional', '', defaultValue='', types=[QgsProcessing.TypeRaster,QgsProcessing.TypeVectorPoint])" ) );
   code = def->asScriptCode();
   QCOMPARE( code, QStringLiteral( "##non_optional=layer raster point" ) );
+  def->setDataTypes( QList< int >() << QgsProcessing::TypePlugin );
+  pythonCode = def->asPythonString();
+  QCOMPARE( pythonCode, QStringLiteral( "QgsProcessingParameterMapLayer('non_optional', '', defaultValue='', types=[QgsProcessing.TypePlugin])" ) );
+  code = def->asScriptCode();
+  QCOMPARE( code, QStringLiteral( "##non_optional=layer plugin" ) );
```
Comment (nyalldawson) [inline tests/src/analysis/testqgsprocessing.cpp:2884]: There's no tests for QgsProcessingUtils::compatiblePluginLayers yet

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp [hunk 3]
```diff
@@ -404,6 +405,17 @@ void QgsProcessingMultipleInputPanelWidget::populateFromProject( QgsProject *pro
       break;
     }
 
+    case QgsProcessing::TypePlugin:
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:408]: Would be good to extend TestProcessingGui::testMultipleFileSelectionDialog() to cover this too

**Review by nyalldawson — COMMENTED** [review]
Looking good!

File: src/core/processing/qgsprocessingutils.cpp (hunk 1 above)
Comment (alexbruy) [reply src/core/processing/qgsprocessingutils.cpp:74]: Added templated private `compatibleMapLayers` and then use it inside compatibleRasterLayers/compatibleMeshLayers/compatiblePluginLayers.

File: tests/src/analysis/testqgsprocessing.cpp (hunk 2 above)
Comment (alexbruy) [reply tests/src/analysis/testqgsprocessing.cpp:2951]: Done

File: src/gui/processing/qgsprocessingmultipleselectiondialog.cpp (hunk 3 above)
Comment (alexbruy) [reply src/gui/processing/qgsprocessingmultipleselectiondialog.cpp:408]: Done.

File: src/core/processing/qgsprocessingutils.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/processing/qgsprocessingutils.cpp:74]: looks great!

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**